### DISCUSSION TITLE
Implicit OPENSSL_init_crypto calls and library contexts

### DISCUSSION THREAD

**Discussion by ifranzki** [openssl/openssl#21864] [Q&A] [answered]
I recently run into a strange situation with OpenSSL's implicit initialization in combination with library contexts:

OpenSSL does perform implicit initialization at various places. I.e. an call to `EVP_PKEY_CTX_new_xxx()` to initialize a PKEY context does internally call `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL)` (`EVP_PKEY_CTX_new_xxx()` -> `int_ctx_new()` -> `ENGINE_get_pkey_meth_engine()` -> `ossl_engine_table_select()` -> `OPENSSL_init_crypto()`).

This causes the OpenSSL configuration to be read (if not already done) and to load all provider (and engines) configured there.  This certainly makes sense in general.

However, in my case I am in a library that gets called from an application that does not itself use OpenSSL. So the application does not explicitly initialize OpenSSL. 

My library sets up its own library context with only the default provider loaded (to not tamper with the applications library context, and to ensure it uses the default provider) and sets the own library context as the default one via `OSSL_LIB_CTX_set0_default()` on entry to my library, and restores the previous one on exit from my library. 

Strangely, if the calling application did not initialize OpenSSL before, then the first call to `EVP_PKEY_CTX_new_xxx()` done by my library will read the OpenSSL config and load the configured providers into the currently active default library context. `OPENSSL_init_crypto()` essentially calls `CONF_modules_load_file_ex(NULL, ...)` with NULL as library context, thus using the current default one. Since my library has set up its own library context and made it the default one (as long as it is within my library), the configured providers gets loaded into my own library context, not the system default one. 

Furthermore, since `OPENSSL_init_crypto()` uses RUN_ONCE to perform the configuration load, this will never ever happen again. So now, the configured providers are loaded in my own library context (which I don't want), and there is no chance to get them loaded into the system default library context (which might be wanted by other libraries or the application).

Sure, I can solves this be calling `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL)` explicitly in my library constructor before setting up my own library context.

However, the question remains: Should `OPENSSL_init_crypto()` load the providers into the current default library context, or should it always load it into the system default library context (i.e. `default_context_int`)? 

The way it is currently, the point in time where `OPENSSL_init_crypto()` is called matters, and may lead to (at least) unexpected behavior. And since initialization may/will  happen implicit within other OpenSSL API calls, it very difficult to predict what happens, and into which library context the providers are loaded.

So loading the configured providers always into the system default library context (i.e. `default_context_int`), regardless which one is currently set as default one, would make things much more consistent.

Another way of solving this inconsistency would be to implicitly initialize OpenSSL also within `OSSL_LIB_CTX_set0_default()`, before it actually sets the new default context. 

Any opinions on this?

**Comment by mattcaswell** [comment, marked as answer]
> Strangely, if the calling application did not initialize OpenSSL before, then the first call to EVP_PKEY_CTX_new_xxx() done by my library will read the OpenSSL config and load the configured providers into the currently active default library context. OPENSSL_init_crypto() essentially calls CONF_modules_load_file_ex(NULL, ...) with NULL as library context, thus using the current default one. Since my library has set up its own library context and made it the default one (as long as it is within my library), the configured providers gets loaded into my own library context, not the system default one.

This sounds like an OpenSSL bug. I think it should always load into the original default ctx.

**Reply by ifranzki** [reply]
A possible fix would be this:

```
diff --git a/crypto/conf/conf_sap.c b/crypto/conf/conf_sap.c
index be87aaf7e8..2e18488664 100644
--- a/crypto/conf/conf_sap.c
+++ b/crypto/conf/conf_sap.c
@@ -65,7 +65,8 @@ int ossl_config_int(const OPENSSL_INIT_SETTINGS *settings)
 #endif
 
 #ifndef OPENSSL_SYS_UEFI
-    ret = CONF_modules_load_file(filename, appname, flags);
+    ret = CONF_modules_load_file_ex(OSSL_LIB_CTX_get0_global_default(),
+                                    filename, appname, flags);
 #else
     ret = 1;
 #endif
```

I can submit a PR if you agree to this fix.

**Reply by mattcaswell** [reply]
This looks reasonable to me. If you don't mind doing it - please raise the PR!

**Reply by ifranzki** [reply]
See https://github.com/openssl/openssl/pull/21897

**Reply by paulidale** [reply]
Hmmmm....  I don't agree that we should override what the user has set and initialise a different default library context.  The user has called `OSSL_LIB_CTX_set0_default()` deliberately.  For all intents and purposes, that context should replace the old default context.  That is, I would expect this to load the configured defaults into `myctx`:
```c
OSSL_LIB_CTX_set0_default(myctx);
OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL);
```

The `OSSL_LIB_CTX_set0_default()` call is a kludge put in because we don't yet pass library contexts around pervasively.  I would strongly recommend **not** using it wherever possible.

**Reply by mattcaswell** [reply]
> The user has called OSSL_LIB_CTX_set0_default() deliberately. For all intents and purposes, that context should replace the old default context.

I completely disagree with this. OPENSSL_init_crypto() should only ever initialise the global default context and never any other lib ctx. Whether OPENSSL_init_crypto() actually kicks in and initialises anything is driven be global state. `OSSL_LIB_CTX_set0_default()` is thread local. You're going to end up with races between threads competing over which libctx gets initialised if `OSSL_LIB_CTX_set0_default()` impacts this. Also, for a library being called by an application, whether OPENSSL_init_crypto() has already been called by the application or not is non-deterministic. A library cannot know this - and should not need to know.
